src/cross-validate.ts

```diff
@@ -0,0 +1,281 @@
+      error: 0,
+    };
+
+    const stats:
+      | ICrossValidateStatsResultStats
+      | ICrossValidateStatsResultBinaryStats = {
+      total: 0,
+      testSize: 0,
+      trainSize: 0,
+    };
+
+    const binaryStats: ICrossValidateStatsResultBinaryStats = {
+      total: 0,
+      testSize: 0,
+      trainSize: 0,
+      truePos: 0,
+      trueNeg: 0,
+      falsePos: 0,
+      falseNeg: 0,
+      precision: 0,
+      recall: 0,
+      accuracy: 0,
+    };
+
+    const results = [];
+    let isBinary = null;
+
+    for (let i = 0; i < k; i++) {
+      const dclone = data.slice(0);
+      const testSet = dclone.splice(i * size, size);
+      const trainSet = dclone;
+      const result = this.testPartition(trainOpts, trainSet, testSet);
+
+      if (isBinary === null) {
+        isBinary =
+          result.hasOwnProperty('falseNeg') &&
+          result.hasOwnProperty('falsePos') &&
+          result.hasOwnProperty('trueNeg') &&
+          result.hasOwnProperty('truePos');
+        if (isBinary) {
+          Object.assign(stats, binaryStats);
+        }
+      }
+
+      avgs.iterations += result.iterations;
+      avgs.testTime += result.testTime;
+      avgs.trainTime += result.trainTime;
+      avgs.error += result.error;
+      stats.total += result.total;
+      if (
+        CrossValidate.isBinaryStats(stats) &&
+        CrossValidate.isBinaryPartitionResults(result)
+      ) {
+        stats.accuracy += result.accuracy;
+        stats.falseNeg += result.falseNeg;
+        stats.falsePos += result.falsePos;
+        stats.precision += result.precision;
+        stats.recall += result.recall;
+        stats.trueNeg += result.trueNeg;
+        stats.truePos += result.truePos;
+      }
+
+      results.push(result);
+    }
+    avgs.error /= k;
+    avgs.iterations /= k;
+    avgs.testTime /= k;
+    avgs.trainTime /= k;
+
+    if (CrossValidate.isBinaryStats(stats)) {
+      stats.precision = stats.truePos / (stats.truePos + stats.falsePos);
+      stats.recall = stats.truePos / (stats.truePos + stats.falseNeg);
+      stats.accuracy = (stats.trueNeg + stats.truePos) / stats.total;
+    }
+
+    stats.testSize = size;
+    stats.trainSize = data.length - size;
+
+    this.json = {
+      avgs: avgs,
+      stats: stats,
+      sets: results,
+    };
+    return this.json;
+  }
+
+  toNeuralNetwork(): NeuralNetwork {
+    if (this.json == null) {
+      throw new Error(
+        'CV json is null, you must call .train() first before getting the neural network'
+      );
+    }
+    return this.fromJSON(this.json);
+  }
+
+  toJSON(): ICrossValidateJSON | null {
+    return this.json;
+  }
+
+  fromJSON(crossValidateJson: ICrossValidateJSON): NeuralNetwork {
+    const Classifier = this.Classifier;
+    const json:
+      | ICrossValidationTestPartitionResults
+      | ICrossValidationTestPartitionBinaryResults = (crossValidateJson as ICrossValidateStats).sets.reduce(
+      (prev, cur) => (prev.error < cur.error ? prev : cur)
+    );
```

Comment:
lines 271-273 I expected to work without needing to assert crossValidateJson as one of the stats or binary stats type but typescript kept giving an error and said something about the .reduce function being incompatible or something. Open to better solutions to this.